src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs

```diff
@@ -355,6 +355,529 @@ public class b
+        }
+
+        [Fact]
+        public void TestMemberNamesReused_UnaffectedByTypeNameChange()
+        {
+            var firstTree = SyntaxFactory.ParseSyntaxTree(@"
+using System;
+
+namespace N1
+{
+    namespace N2.N3
+    {
+        class C
+        {
+            int x, y, z;
+        }
+    }
+}
+");
+            var secondTree = SyntaxFactory.ParseSyntaxTree(@"
+using System;
+
+namespace N1
+{
+    namespace N2.N3
+    {
+        class D
+        {
+            int z, y, x;
+        }
+    }
+}
+");
+
+            var compilation = CreateCompilation(new SyntaxTree[] { firstTree });
+
+            var type1 = (SourceNamedTypeSymbol)compilation.GetTypeByMetadataName("N1.N2.N3.C");
+            Assert.True(type1.MergedDeclaration.Declarations[0].MemberNames.Value.SetEquals(new[] { "x", "y", "z" }));
+
+            compilation = compilation.ReplaceSyntaxTree(firstTree, secondTree);
+
+            var type2 = (SourceNamedTypeSymbol)compilation.GetTypeByMetadataName("N1.N2.N3.D");
+            Assert.True(type2.MergedDeclaration.Declarations[0].MemberNames.Value.SetEquals(new[] { "x", "y", "z" }));
+
+            // We should have the exact same set for the names.
+            Assert.Same(type1.MergedDeclaration.Declarations[0].MemberNames, type2.MergedDeclaration.Declarations[0].MemberNames);
+        }
+
+        [Fact]
+        public void TestMemberNamesReused_UnaffectedByNamespaceChange()
+        {
+            var firstTree = SyntaxFactory.ParseSyntaxTree(@"
+using System;
+
+namespace N1
+{
+    namespace N2.N3
+    {
+        class C
+        {
+            int x, y, z;
+        }
+    }
+}
+");
+            var secondTree = SyntaxFactory.ParseSyntaxTree(@"
+using System;
+
+namespace N1
+{
+    class C
+    {
+        int z, y, x;
+    }
+}
+");
+
+            var compilation = CreateCompilation(new SyntaxTree[] { firstTree });
+
+            var type1 = (SourceNamedTypeSymbol)compilation.GetTypeByMetadataName("N1.N2.N3.C");
+            Assert.True(type1.MergedDeclaration.Declarations[0].MemberNames.Value.SetEquals(new[] { "x", "y", "z" }));
+
+            compilation = compilation.ReplaceSyntaxTree(firstTree, secondTree);
+
+            var type2 = (SourceNamedTypeSymbol)compilation.GetTypeByMetadataName("N1.C");
+            Assert.True(type2.MergedDeclaration.Declarations[0].MemberNames.Value.SetEquals(new[] { "x", "y", "z" }));
+
+            // We should have the exact same set for the names.
+            Assert.Same(type1.MergedDeclaration.Declarations[0].MemberNames, type2.MergedDeclaration.Declarations[0].MemberNames);
+        }
+
+        [Fact]
+        public void TestMemberNamesReused_SupportsMultipleTypes()
+        {
+            var firstTree = SyntaxFactory.ParseSyntaxTree(@"
+using System;
+
+namespace N1
+{
+    namespace N2.N3
+    {
+        class C
+        {
+            int x, y, z;
+        }
+
+        class D
```

Comment:
What's the expected behavior when, for example, `class D` is moved to be before `class C` in this file?

Reply:
Taht it will not reuse the member names.  As that's a rare occurrence (non-name changing edits within a method body are the common case), its' fine that we simply have the same cost as today (We allocate a new collection).